service/matching/matching_engine.go

```diff
@@ -1328,6 +1328,12 @@ func (e *matchingEngineImpl) DispatchNexusTask(ctx context.Context, request *mat
 		return nil, serviceerror.NewFailedPrecondition("versioning is disabled")
 	}
 
+	if deadline, set := ctx.Deadline(); set {
+		// Update the Request-Timeout header to subtract processing time till now so the worker that receives this
+		// task knows how much time it has left to complete it.
+		request.Request.Header["Request-Timeout"] = time.Until(deadline).String()
+	}
```

Comment:
This shouldn't be here, put the deadline on the request just before returning the task to a poller.